### ISSUE TITLE
stale package bug

### ISSUE THREAD

**Issue by mat813** [freebsd/poudriere#517]
This is with poudriere-devel-3.1.99.20170721.

I noticed some ports being rebuilt and rebuilt time and time again, and began to search what was happening.

To reproduce, `poudriere options -c www/nginx` and check BROTLI (it is in the 3rd party section, after all the HTTP_* options)

With a clean repository where I have neither www/nginx, nor archivers/brotli, if I run:

```
# poudriere bulk -t -N -j 10amd64 -p ports -J 8:40 www/nginx
[00:00:00] ====>> Creating the reference jail... done
[00:00:01] ====>> Mounting system devices for 10amd64-ports
[00:00:01] ====>> Mounting ports/packages/distfiles
[00:00:01] ====>> Stashing existing package repository
[00:00:01] ====>> Mounting ccache from: /p/data/ccache
[00:00:01] ====>> Mounting packages from: /p/data/packages/10amd64-ports
[00:00:01] ====>> Copying /var/db/ports from: /usr/local/etc/poudriere.d/options
[00:00:01] ====>> Appending to make.conf: /usr/local/etc/poudriere.d/make.conf
/etc/resolv.conf -> /p/data/.m/10amd64-ports/ref/etc/resolv.conf
[00:00:01] ====>> Starting jail 10amd64-ports
[00:00:01] ====>> Logs: /p/data/logs/bulk/10amd64-ports/20170727T165043Z
[00:00:01] ====>> WWW: http://aragorn.in.absolight.net/p/build.html?mastername=10amd64-ports&build=20170727T165043Z
[00:00:01] ====>> Loading MOVED
[00:00:01] ====>> Ports supports: DEPENDS_ARGS
[00:00:01] ====>> Gathering ports metadata
[00:00:01] ====>> Calculating ports order and dependencies
[00:00:02] ====>> Sanity checking the repository
[00:00:02] ====>> Checking packages for incremental rebuild needs
[00:00:03] ====>> Deleting stale symlinks... done
[00:00:03] ====>> Deleting empty directories... done
[00:00:03] ====>> Cleaning the build queue
[00:00:03] ====>> Sanity checking build queue
[00:00:03] ====>> Processing PRIORITY_BOOST
[00:00:03] ====>> Balancing pool
[00:00:03] ====>> Recording filesystem state for prepkg... done
[00:00:03] ====>> Building 2 packages using 2 builders
[00:00:03] ====>> Starting/Cloning builders
[00:00:04] ====>> Hit CTRL+t at any time to see build progress and stats
[00:00:04] ====>> [01][00:00:00] Building archivers/brotli | brotli-0.6.0,1
[00:00:10] ====>> [01][00:00:06] Finished archivers/brotli | brotli-0.6.0,1: Success
[00:00:10] ====>> [01][00:00:00] Building www/nginx | nginx-1.12.1,2
[00:00:23] ====>> [01][00:00:13] Finished www/nginx | nginx-1.12.1,2: Success
[00:00:23] ====>> Stopping 2 builders
10amd64-ports-job-01: removed
10amd64-ports-job-02: removed
10amd64-ports-job-01-n: removed
10amd64-ports-job-02-n: removed
[00:00:24] ====>> Committing packages to repository
[00:00:24] ====>> Removing old packages
[00:00:24] ====>> Built ports: archivers/brotli www/nginx
[10amd64-ports] [20170727T165043Z] [committing:] Queued: 2  Built: 2  Failed: 0  Skipped: 0  Ignored: 0  Tobuild: 0   Time: 00:00:23
[00:00:24] ====>> Logs: /p/data/logs/bulk/10amd64-ports/20170727T165043Z
[00:00:24] ====>> WWW: http://aragorn.in.absolight.net/p/build.html?mastername=10amd64-ports&build=20170727T165043Z
[00:00:24] ====>> Cleaning up
10amd64-ports: removed
10amd64-ports-n: removed
[00:00:24] ====>> Unmounting file systems
```

It builds both ports, and all goes well.

Now, if I run the same command, I would expect it to not build anything, but...

```
# poudriere bulk -t -N -j 10amd64 -p ports -J 8:40 www/nginx
[00:00:00] ====>> Creating the reference jail... done
[00:00:01] ====>> Mounting system devices for 10amd64-ports
[00:00:01] ====>> Mounting ports/packages/distfiles
[00:00:01] ====>> Stashing existing package repository
[00:00:01] ====>> Mounting ccache from: /p/data/ccache
[00:00:01] ====>> Mounting packages from: /p/data/packages/10amd64-ports
[00:00:01] ====>> Copying /var/db/ports from: /usr/local/etc/poudriere.d/options
[00:00:01] ====>> Appending to make.conf: /usr/local/etc/poudriere.d/make.conf
/etc/resolv.conf -> /p/data/.m/10amd64-ports/ref/etc/resolv.conf
[00:00:01] ====>> Starting jail 10amd64-ports
[00:00:01] ====>> Logs: /p/data/logs/bulk/10amd64-ports/20170727T165240Z
[00:00:01] ====>> WWW: http://aragorn.in.absolight.net/p/build.html?mastername=10amd64-ports&build=20170727T165240Z
[00:00:01] ====>> Loading MOVED
[00:00:01] ====>> Ports supports: DEPENDS_ARGS
[00:00:01] ====>> Gathering ports metadata
[00:00:01] ====>> Calculating ports order and dependencies
[00:00:02] ====>> Sanity checking the repository
[00:00:02] ====>> Checking packages for incremental rebuild needs
[00:00:02] ====>> Deleting brotli-0.6.0,1.txz: stale package: unwanted origin archivers/py-brotli
[00:00:02] ====>> Deleting nginx-1.12.1,2.txz: new dependency: archivers/brotli
[00:00:02] ====>> Deleting stale symlinks... done
[00:00:02] ====>> Deleting empty directories... done
[00:00:02] ====>> Cleaning the build queue
[00:00:03] ====>> Sanity checking build queue
[00:00:03] ====>> Processing PRIORITY_BOOST
[00:00:03] ====>> Balancing pool
[00:00:03] ====>> Recording filesystem state for prepkg... done
[00:00:03] ====>> Building 2 packages using 2 builders
[00:00:03] ====>> Starting/Cloning builders
[00:00:04] ====>> Hit CTRL+t at any time to see build progress and stats
[00:00:04] ====>> [01][00:00:00] Building archivers/brotli | brotli-0.6.0,1
[00:00:10] ====>> [01][00:00:06] Finished archivers/brotli | brotli-0.6.0,1: Success
[00:00:10] ====>> [01][00:00:00] Building www/nginx | nginx-1.12.1,2
[00:00:23] ====>> [01][00:00:13] Finished www/nginx | nginx-1.12.1,2: Success
[00:00:23] ====>> Stopping 2 builders
10amd64-ports-job-01: removed
10amd64-ports-job-02: removed
10amd64-ports-job-01-n: removed
10amd64-ports-job-02-n: removed
[00:00:24] ====>> Committing packages to repository
[00:00:24] ====>> Removing old packages
[00:00:24] ====>> Built ports: archivers/brotli www/nginx
[10amd64-ports] [20170727T165240Z] [committing:] Queued: 2  Built: 2  Failed: 0  Skipped: 0  Ignored: 0  Tobuild: 0   Time: 00:00:23
[00:00:24] ====>> Logs: /p/data/logs/bulk/10amd64-ports/20170727T165240Z
[00:00:24] ====>> WWW: http://aragorn.in.absolight.net/p/build.html?mastername=10amd64-ports&build=20170727T165240Z
[00:00:24] ====>> Cleaning up
10amd64-ports: removed
10amd64-ports-n: removed
[00:00:24] ====>> Unmounting file systems
```

It deletes brotli saying `Deleting brotli-0.6.0,1.txz: stale package: unwanted origin archivers/py-brotli` I went and check, this one is called pyXX-brotli, not brotli, and if I run the bulk with -i, the brotli package has the right origin:

```
root@10amd64-ports:~ # pkg info brotli | grep Origin
Origin         : archivers/brotli
```

So I don't really know wtf is happening, but something is fishy.

**Comment by bdrewery** [comment]
It's a (still current as of r447456) bug in _MOVED_:
```
# grep brotli MOVED
archivers/brotli|archivers/py-brotli|2016-11-24|Brotli is a python module
devel/libbrotli|archivers/brotli|2017-07-12|meta project no longer required to build libs
```

**Comment by bdrewery** [comment]
I can add an assertion that _MOVED_ cannot contain an old directory that now exists, and/or ignore such entries.

**Comment by bdrewery** [comment]
After fixing Tools/scripts/MOVEDlint.awk I get:
```
8774: archivers/brotli must be marked as resurrected
8934: www/rubygem-jquery-rails must be marked as resurrected
```

**Comment by mat813** [comment]
Damn I tried so many things and did not think of MOVED.

**Comment by bdrewery** [comment]
I cannot recreate this issue anymore (even with restoring MOVED). I'm closing for now until we see another case.

**Comment by hartzell** [comment]
I'm seeing what I think is the same thing, building emacs with the things set to use the `nox` flavor.

After running a complete and happy build of a handful of things and not changing anything, it rebuilds emacs:

```
freebsd@pkg-builder:~ % sudo poudriere bulk -j 11_2 -p default -f /usr/local/etc/poudriere.d/packages
[00:00:00] Creating the reference jail... done
[00:00:00] Mounting system devices for 11_2-default
args=mount /poudriere/data/.m/11_2-default/ref
[00:00:00] Mounting ports/packages/distfiles
[00:00:00] Stashing existing package repository
[00:00:00] Mounting packages from: /poudriere/data/packages/11_2-default
[00:00:00] Appending to make.conf: /usr/local/etc/poudriere.d/make.conf
/etc/resolv.conf -> /poudriere/data/.m/11_2-default/ref/etc/resolv.conf
[00:00:00] Starting jail 11_2-default
args=start
[00:00:01] Logs: /poudriere/data/logs/bulk/11_2-default/2019-05-05_21h51m50s
[00:00:01] Loading MOVED for /poudriere/data/.m/11_2-default/ref/usr/ports
[00:00:03] Ports supports: FLAVORS SELECTED_OPTIONS
[00:00:03] Gathering ports metadata
[00:00:05] Calculating ports order and dependencies
[00:00:05] Sanity checking the repository
[00:00:05] Checking packages for incremental rebuild needs
[00:00:05] Deleting emacs-nox-26.2,3.txz: stale package: unwanted origin editors/emacs@nox
[00:00:06] Deleting stale symlinks... done
[00:00:06] Deleting empty directories... done
[00:00:06] Cleaning the build queue
[00:00:06] Sanity checking build queue
[00:00:06] Processing PRIORITY_BOOST
[00:00:06] Balancing pool
[00:00:06] Recording filesystem state for prepkg... done
[00:00:07] Building 1 packages using 1 builders
[00:00:07] Starting/Cloning builders
args=stop 01 /poudriere/data/.m/11_2-default/01
args=mount /poudriere/data/.m/11_2-default/01
args=start 01 /poudriere/data/.m/11_2-default/01
[00:00:07] Hit CTRL+t at any time to see build progress and stats
[00:00:07] [01] [00:00:00] Building editors/emacs@nox | emacs-nox-26.2,3
[00:02:57] [01] [00:02:50] Finished editors/emacs@nox | emacs-nox-26.2,3: Success
[00:02:58] Stopping 1 builders
args=stop 01 /poudriere/data/.m/11_2-default/01
[00:02:58] Creating pkg repository
Creating repository in /tmp/packages: 100%
Packing files for repository: 100%
[00:03:00] Committing packages to repository: /poudriere/data/packages/11_2-default/.real_1557093290 via .latest symlink
[00:03:00] Removing old packages
[00:03:00] Built ports: editors/emacs
[11_2-default] [2019-05-05_21h51m50s] [committing:] Queued: 1  Built: 1  Failed: 0  Skipped: 0  Ignored: 0  Tobuild: 0   Time: 00:02:59
[00:03:00] Logs: /poudriere/data/logs/bulk/11_2-default/2019-05-05_21h51m50s
[00:03:00] Cleaning up
args=stop
[00:03:00] Unmounting file systems
freebsd@pkg-builder:~ %
```

I'm using a system that's a mashup of [these DigitalOcean instructions](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-poudriere-build-system-to-create-packages-for-your-freebsd-servers) and [this Ansible role](https://github.com/reallyenglish/ansible-role-poudriere).

Here's the relevant bit of my `make.conf`:

```
# The default flavor of emacs is nox, and other ports should
# depend on nox.
.if ${.CURDIR:M*/editors/emacs*}
FLAVOR?=nox
.endif
DEFAULT_VERSIONS+=emacs=nox
```

Here's my sandbox `packages` file:

```
editors/emacs
sysutils/htop
sysutils/tmux
sysutils/tree
textproc/jq
textproc/the_silver_searcher
```

What more can I tell you?

**Comment by hartzell** [comment]
Poking around at this a bit, I think I see what's going on.

I've configured my world, via the `make.conf` snippet above, so that if/when I ask to install `emacs` I install `emacs@nox` instead.   When `pourdriere` is assembling/cleansing its list of ports, it see's that I've requested `emacs` and/but finds that there's an apparently unwanted `emacs@nox` package installed so it cleans it up.

Seems like the fragile bit is *"The first flavor is the default one [...]"* (from the [flavors docs](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/flavors-using.html)).  It might be useful to have an explicit way to configure the default flavor and then things like pourdiere could more easily check for it.

I think that the important part of my `make.conf` config is the bit that sets adds `emacs=nox` to the list of `DEFAULT_VERSIONS`, which prevents other ports from unexpectedly/unwantedly building an X11 `emacs`.

As long as I explicitly ask for `emacs@nox` in my list of ports to build, things seem to work.

**Comment by hartzell** [comment]
Poking around at this a bit, I think I see what's going on.

I've configured my world, via the `make.conf` snippet above, so that if/when I ask to install `emacs` I install `emacs@nox` instead.   When `pourdriere` is assembling/cleansing its list of ports, it see's that I've requested `emacs` and/but finds that there's an apparently unwanted `emacs@nox` package installed so it cleans it up.

Seems like the fragile bit is *"The first flavor is the default one [...]"* (from the [flavors docs](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/flavors-using.html)).  It might be useful to have an explicit way to configure the default flavor and then things like pourdiere could more easily check for it.

I think that the important part of my `make.conf` config is the bit that sets adds `emacs=nox` to the list of `DEFAULT_VERSIONS`, which prevents other ports from unexpectedly/unwantedly building an X11 `emacs`.

As long as I explicitly ask for `emacs@nox` in my list of ports to build, things seem to work.
